sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java

```diff
@@ -39,26 +40,24 @@
         List<CredentialUnavailableException> exceptions = new ArrayList<>(4);
         return Flux.fromIterable(credentials)
                    .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {
+                       if (!t.getClass().getSimpleName().equals("CredentialUnavailableException")) {
+                           return Mono.error(new CredentialUnavailableException(
+                                   unavailableError + p.getClass().getSimpleName() + " authentication failed.",
+                                   t));
+                       }
                        exceptions.add(t);
                        return Mono.empty();
                    }), 1)
                    .next()
                    .switchIfEmpty(Mono.defer(() -> {
-
-                       StringBuilder message = new StringBuilder("Tried "
-                             + credentials.stream().map(c -> c.getClass().getSimpleName())
-                                   .collect(Collectors.joining(", "))
-                             + " but failed to acquire a token for any of them. Please verify the"
-                             + " environment for the credentials"
-                             + " and see more details in the causes below.");
-
                        // Chain Exceptions.
                        CredentialUnavailableException last = exceptions.get(exceptions.size() - 1);
                        for (int z = exceptions.size() - 2; z >= 0; z--) {
                            CredentialUnavailableException current = exceptions.get(z);
-                           last = new CredentialUnavailableException(current.getMessage(), last);
+                           last = new CredentialUnavailableException(current.getMessage() + " " + last.getMessage(),
+                                last.getCause());
                        }
-                       return Mono.error(new CredentialUnavailableException(message.toString(), last));
+                       return Mono.error(last);
```

Comment:
WHat does the error message with stack trace look like now ?
Can you post the sample output here ?

Reply:
Exception in thread "main" java.lang.RuntimeException: Max retries 3 times exceeded. Error Details: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.
ManagedIdentityCredential authentication unavailable. Connection to IMDS endpoint cannot be established, Network is unreachable: connect.
SharedTokenCacheCredential authentication unavailable. No accounts were found in the cache.
IntelliJ Authentication not available. Please log in with Azure Tools for IntelliJ plugin in the IDE.
Failed to read Vs Code credentials from Windows Credential API.
AzureCliCredential authentication unavailable. Azure CLI not installed
        at com.azure.core.http.policy.RetryPolicy.lambda$attemptAsync$1(RetryPolicy.java:119)
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:88)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)
        at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)
        at reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:251)
        at reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:251)
        at reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onError(FluxPeekFuseable.java:894)
        at reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onError(FluxPeekFuseable.java:894)
        at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onError(Operators.java:1994)
        at reactor.core.publisher.Operators.error(Operators.java:196)
        at reactor.core.publisher.MonoError.subscribe(MonoError.java:52)
        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:52)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4218)
        at reactor.core.publisher.FluxSwitchIfEmpty$SwitchIfEmptySubscriber.onComplete(FluxSwitchIfEmpty.java:75)
        at reactor.core.publisher.MonoNext$NextSubscriber.onComplete(MonoNext.java:96)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:838)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:600)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.innerComplete(FluxFlatMap.java:909)
        at reactor.core.publisher.FluxFlatMap$FlatMapInner.onComplete(FluxFlatMap.java:1013)
        at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onComplete(Operators.java:1989)
        at reactor.core.publisher.Operators.complete(Operators.java:135)
        at reactor.core.publisher.MonoEmpty.subscribe(MonoEmpty.java:45)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4218)
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:97)
        at reactor.core.publisher.FluxMap$MapSubscriber.onError(FluxMap.java:126)
        at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onError(Operators.java:1994)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:135)
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:73)
        at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1755)
        at reactor.core.publisher.MonoCompletionStage.lambda$subscribe$0(MonoCompletionStage.java:86)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
        Suppressed: java.lang.Exception: #block terminated with an error
                at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
                at reactor.core.publisher.Mono.block(Mono.java:1678)
                at com.azure.security.keyvault.secrets.SecretClient.setSecretWithResponse(SecretClient.java:113)
                at com.azure.security.keyvault.secrets.SecretClient.setSecret(SecretClient.java:93)
                at com.azure.identity.Test.main(Test.java:31)
Caused by: com.azure.identity.CredentialUnavailableException: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.
ManagedIdentityCredential authentication unavailable. Connection to IMDS endpoint cannot be established, Network is unreachable: connect.
SharedTokenCacheCredential authentication unavailable. No accounts were found in the cache.
IntelliJ Authentication not available. Please log in with Azure Tools for IntelliJ plugin in the IDE.
Failed to read Vs Code credentials from Windows Credential API.
AzureCliCredential authentication unavailable. Azure CLI not installed
        at com.azure.identity.ChainedTokenCredential.lambda$2(ChainedTokenCredential.java:58)
        at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:44)
        ... 28 more

Reply:
if failed：
Exception in thread "main" java.lang.RuntimeException: Max retries 3 times exceeded. Error Details: DefaultAzureCredential authentication failed. ---> EnvironmentCredential authentication failed. Error Details:AADSTS90002: Tenant '72f988bf-86f1-41af-91ab-2d7cd011db46' not found. This may happen if there are no active subscriptions for the tenant. Check to
make sure you have the correct tenant ID. Check with your subscription administrator.
Trace ID: 43ffc9ad-f468-4665-ac3a-14a278a98600
Correlation ID: e3d3e754-89f0-4839-870c-57e6aba9a5f0
Timestamp: 2020-06-08 01:49:43Z
        at com.azure.core.http.policy.RetryPolicy.lambda$attemptAsync$1(RetryPolicy.java:119)
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:88)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
        at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)
        at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)
        at reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:251)
        at reactor.core.publisher.MonoPeekTerminal$MonoTerminalPeekSubscriber.onError(MonoPeekTerminal.java:251)
        at reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onError(FluxPeekFuseable.java:894)
        at reactor.core.publisher.FluxPeekFuseable$PeekConditionalSubscriber.onError(FluxPeekFuseable.java:894)
        at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onError(Operators.java:1994)
        at reactor.core.publisher.MonoNext$NextSubscriber.onError(MonoNext.java:87)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.checkTerminated(FluxFlatMap.java:834)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:600)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:580)
        at reactor.core.publisher.FluxFlatMap$FlatMapMain.innerError(FluxFlatMap.java:855)
        at reactor.core.publisher.FluxFlatMap$FlatMapInner.onError(FluxFlatMap.java:1006)
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:100)
        at reactor.core.publisher.Operators.error(Operators.java:196)
        at reactor.core.publisher.MonoError.subscribe(MonoError.java:52)
        at reactor.core.publisher.Mono.subscribe(Mono.java:4218)
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:97)
        at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onError(FluxMapFuseable.java:134)
        at reactor.core.publisher.MonoCompletionStage.lambda$subscribe$0(MonoCompletionStage.java:80)
        at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859)
        at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1705)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1692)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
        Suppressed: java.lang.Exception: #block terminated with an error
                at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
                at reactor.core.publisher.Mono.block(Mono.java:1678)
                at com.azure.security.keyvault.secrets.SecretClient.setSecretWithResponse(SecretClient.java:113)
                at com.azure.security.keyvault.secrets.SecretClient.setSecret(SecretClient.java:93)
                at com.azure.identity.Test.main(Test.java:31)
Caused by: com.azure.identity.CredentialUnavailableException: DefaultAzureCredential authentication failed. ---> EnvironmentCredential authentication failed. Error Details:AADSTS90002: Tenant '72f988bf-86f1-41af-91ab-2d7cd011db46' not found. This may happen if there are no active subscriptions for the tenant. Check to make sure you have the correct tenant
ID. Check with your subscription administrator.
Trace ID: 43ffc9ad-f468-4665-ac3a-14a278a98600
Correlation ID: e3d3e754-89f0-4839-870c-57e6aba9a5f0
Timestamp: 2020-06-08 01:49:43Z
        at com.azure.identity.ChainedTokenCredential.lambda$1(ChainedTokenCredential.java:44)
        at reactor.core.publisher.Mono.lambda$onErrorResume$31(Mono.java:3360)
        at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:88)
        ... 12 more
Caused by: com.microsoft.aad.msal4j.MsalServiceException: AADSTS90002: Tenant '72f988bf-86f1-41af-91ab-2d7cd011db46' not found. This may happen if there are no active subscriptions for the tenant. Check to make sure you have the correct tenant ID. Check with your subscription administrator.
Trace ID: 43ffc9ad-f468-4665-ac3a-14a278a98600
Correlation ID: e3d3e754-89f0-4839-870c-57e6aba9a5f0
Timestamp: 2020-06-08 01:49:43Z
        at com.microsoft.aad.msal4j.MsalServiceExceptionFactory.fromHttpResponse(MsalServiceExceptionFactory.java:43)
        at com.microsoft.aad.msal4j.TokenRequestExecutor.createAuthenticationResultFromOauthHttpResponse(TokenRequestExecutor.java:81)
        at com.microsoft.aad.msal4j.TokenRequestExecutor.executeTokenRequest(TokenRequestExecutor.java:36)
        at com.microsoft.aad.msal4j.ClientApplicationBase.acquireTokenCommon(ClientApplicationBase.java:92)
        at com.microsoft.aad.msal4j.AcquireTokenByAuthorizationGrantSupplier.execute(AcquireTokenByAuthorizationGrantSupplier.java:52)
        at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:59)
        at com.microsoft.aad.msal4j.AuthenticationResultSupplier.get(AuthenticationResultSupplier.java:17)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
        ... 6 more